Python/codecs.c

```diff
@@ -774,50 +774,74 @@ PyObject *PyCodec_IgnoreErrors(PyObject *exc)
 }
 
 
-PyObject *PyCodec_ReplaceErrors(PyObject *exc)
+// --- handler: 'replace' -----------------------------------------------------
+
+static PyObject *
+_PyCodec_ReplaceUnicodeEncodeError(PyObject *exc)
 {
     Py_ssize_t start, end, slen;
+    if (_PyUnicodeError_GetParams(exc, NULL, NULL,
+                                  &start, &end, &slen, false) < 0)
+    {
+        return NULL;
+    }
+    PyObject *res = PyUnicode_New(slen, '?');
+    if (res == NULL) {
+        return NULL;
+    }
+    assert(PyUnicode_KIND(res) == PyUnicode_1BYTE_KIND);
+    Py_UCS1 *outp = PyUnicode_1BYTE_DATA(res);
+    memset(outp, '?', sizeof(Py_UCS1) * slen);
+    assert(_PyUnicode_CheckConsistency(res, 1));
+    return Py_BuildValue("(Nn)", res, end);
+}
 
-    if (PyObject_TypeCheck(exc, (PyTypeObject *)PyExc_UnicodeEncodeError)) {
-        if (_PyUnicodeError_GetParams(exc, NULL, NULL,
-                                      &start, &end, &slen, false) < 0) {
-            return NULL;
-        }
-        PyObject *res = PyUnicode_New(slen, '?');
-        if (res == NULL) {
-            return NULL;
-        }
-        assert(PyUnicode_KIND(res) == PyUnicode_1BYTE_KIND);
-        Py_UCS1 *outp = PyUnicode_1BYTE_DATA(res);
-        memset(outp, '?', sizeof(Py_UCS1) * slen);
-        assert(_PyUnicode_CheckConsistency(res, 1));
-        return Py_BuildValue("(Nn)", res, end);
+
+static PyObject *
+_PyCodec_ReplaceUnicodeDecodeError(PyObject *exc)
+{
+    Py_ssize_t end;
+    if (PyUnicodeDecodeError_GetEnd(exc, &end) < 0) {
+        return NULL;
     }
-    else if (PyObject_TypeCheck(exc, (PyTypeObject *)PyExc_UnicodeDecodeError)) {
-        if (_PyUnicodeError_GetParams(exc, NULL, NULL,
-                                      NULL, &end, NULL, true) < 0) {
-            return NULL;
-        }
-        return Py_BuildValue("(Cn)",
-                             (int)Py_UNICODE_REPLACEMENT_CHARACTER,
-                             end);
+    // Note: Py_UNICODE_REPLACEMENT_CHARACTER < (2 ** 16) < INT_MAX
```

Comment:
Hm, where do we assert this?

(C only guarantees `(2**15-1) <= INT_MAX`. A bunch of stuff in CPython will currently choke on 16-bit ints, but in this kind of clean-up commit it would be nice to document where the limit is checked, or add a static_assert, or use `_Py_STR` or `u#`.)

Reply:
Py_UNICODE_REPLACEMENT_CHARACTER is actually `?` but we can assert it. I also messed up my bound on INT_MAX :'). But I can assert it here actually.

Reply:
So. 

- Py_UNICODE_REPLACEMENT_CHARACTER is 0xFFFD = $2^16-3$.
- If the largest int is actually $2^15-1$ and not $2^31$, then we're indeed in a bad situation.
- Let's check at build time that `INT_MAX` is sufficiently large? (maybe just >= 2^16 would be sufficient? but I don't know if it's possible to have ints with 24 bits, say on very esoteric architectures that are more toy architectures).

Reply:
Oh dear. You're heading into a rabbit hole; are you sure you want to continue? [y/n]

In this case, looking at the options, I'd recommend switching to `_Py_STR` -- the size is known so we can allocate a `'�'` statically.

Reply:
> Oh dear. You're heading into a rabbit hole; are you sure you want to continue? [y/n]

y... err... n.

Ok for using `_Py_STR` although I never remember how to use as I barely use it :) (today I won't code bc I'm busy but I'll try to do it tomorrow)

Reply:
I hope you don't mind if I do it :)

Reply:
No I don't. In general, you can edit my PR however you want. If I'm not happy with the edit, I usually post-edit it xD but if it can save time, I don't mind someone else doing it (otherwise I would uncheck the box "allow edit")

Reply:
... Oh, never mind. `_Py_STR` is currently ASCII-only.

IMO it would be OK to just remove the comment for now, it's wrong. On a 16-bit system, `PyUnicode_FromOrdinal` will reject the value... but no one complained since 2002.
```suggestion
```